File: array/cpp/RepeatingMissingNumber.cpp
#include <bits/stdc++.h>
using namespace std;

void swap(int *arr,int i, int j){
    int temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;
}

int *repeatingMissingNumberBruteForce(int *arr, int arrLength) {
    
    sort(arr,arr+arrLength);

    int *result=new int[2];

    for(int i=0;i<arrLength;i++){
        if (i == arrLength-1){
            break;
        }

        if(arr[i]==arr[i+1]){
            result[0]=arr[i]; 
            break;
        }
    }

    for(int i=0;i<arrLength;i++){
        if (i == arrLength-1){
            break;
        }

        if(arr[i] != i+1){
            result[1]=i+1; 
            break;
        }
    }


    return result;
}

int *repeatingMissingNumberBetter(int *arr, int arrLength) {
    
    int *result=new int[2];
    int countArray[arrLength]={0};

    for(int i=0;i<arrLength;i++){
        countArray[arr[i]-1]+=1;
    }

    for (int i = 0; i < arrLength; ++i){
        if (countArray[i] == 0){
            result[1] = i+1;
        }
        
        if (countArray[i] == 2){
            result[0] = i+1;
        }
    }
      
    return result;
}

int *repeatingMissingNumberOptimized(int *arr, int arrLength) {
    int *result=new int[2];

    int i = 0;
    while (i < arrLength) {
        if (arr[i] == arr[arr[i] - 1]) {
            i++;
        }
        else {
            if (arr[i] != arr[arr[i] - 1]) {
                swap(arr, i, arr[i] - 1);
            }
            else {
                i++;
            }
        }
    }

    for (i = 0; i < arrLength; i++) {
        if (arr[i] != i+1) {
            result[0] = arr[i];
            result[1] = i+1;
            break;
        }
    }

    return result;
}
int main()
{
    int arr[] = { 3, 2, 3 };
    int arrLength = sizeof(arr) / sizeof(arr[0]);

    cout << "\nBrute-force\n";
    int *result = repeatingMissingNumberBruteForce(arr,arrLength);
    for (int i = 0; i<2;i++){
        cout << result[i] <<"\n";
    }
    cout << "Better\n";
    result = repeatingMissingNumberBetter(arr,arrLength);
    for (int i = 0; i<2;i++){
        cout << result[i] <<"\n";
    }
    
    cout << "Optimized\n";
    result = repeatingMissingNumberOptimized(arr,arrLength);
    for (int i = 0; i<2;i++){
        cout << result[i] <<"\n";
    }
    return 0;
}

File: array/python/RepeatingMissingNumber.py
def repeating_missing_number_brute_force(arr):
    arr_length = len(arr)
    
    arr.sort()
    # print(arr)
    result = [0,0]

    for i in range(arr_length):
        if i == arr_length-1:
            break
        
        if arr[i] == arr[i+1]:
            result[0] = arr[i]
            break

    for i in range(arr_length):
        if i == arr_length-1:
            break

        if arr[i] != i+1:
            result[1] = i+1
            break
    
    return result

def repeating_missing_number_better(arr):
    arr_length = len(arr)

    result = [0,0]

    count_array = [0] * (len(arr))

    for i in range(arr_length):
        count_array[arr[i]-1]+=1
    
    for i in range(arr_length):
        if count_array[i] == 0:
            result[1] = i+1

        if count_array[i] == 2:
            result[0] = i+1

    return result

def repeating_missing_number_optimized(arr):
    arr_length = len(arr)

    result = [0,0]
    
    i = 0

    while (i< arr_length):
        if arr[i] == arr[arr[i]-1]:
            i+=1
        else:
            if arr[i] != arr[arr[i]-1]:
                arr[i],arr[arr[i]-1] = arr[arr[i]-1],arr[i]
            else:
                i+=1
    
    for i in range(arr_length):
        if arr[i] != i+1:
            result[0] = arr[i]
            result[1] = i+1;
            break
    
    return result

arr = [3,1,3]

print("Brute-force")
for i in repeating_missing_number_brute_force(arr):
    print(i)

print("Better")
for i in repeating_missing_number_better(arr):
    print(i)

print("Optimized")
for i in repeating_missing_number_optimized(arr):
    print(i)